Add LeNet example

huningxin opened the pull request:
Fix https://github.com/webmachinelearning/webnn/issues/70

To facilitate the review, a working version is hosted at: https://huningxin.github.io/webnn-samples/lenet/

This example is based on the foundation webnn-polyfill in https://github.com/webmachinelearning/webnn-polyfill/pull/1

@anssiko @wchao1115 @pyu10055, please take a look. Thanks!

anssiko commented:
Thanks looks great! Small nit on narrow screens, e.g. mobile: I suggest you use CSS media queries to position prediction results and controls underneath the image on narrow screens. Currently, they're overlaid on top of the image on mobile, and as such, unreadable.

pyu10055 commented on lenet/index.html:
```diff
@@ -0,0 +1,118 @@
+			display: none;
+		}
+		.radioSelected{
+			display: inline-block;
```
it would good to sort the css attributes alphabetically.

huningxin replied:
good point. I'll fix that.

huningxin replied:
Fixed. Actually I found these css attributes are unused, so I removed them.

pyu10055 commented on lenet/index.html:
```diff
@@ -0,0 +1,118 @@
+    <script src="https://code.jquery.com/jquery-3.5.1.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
+    <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js" integrity="sha384-9/reFTGAW83EW2RDu2S0VKaIzap3H66lZH81PoYlFhbGU+6BZp6G7niu735Sk7lN" crossorigin="anonymous"></script>
+    <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js" integrity="sha384-B4gt1jrGC7Jh4AgTPSdUtOBvfO8shuf57BaghqFfPlYxofvL8/KUEfYiJOMMV+rV" crossorigin="anonymous"></script>
+    <script src="https://webmachinelearning.github.io/webnn-polyfill/dist/webnn-polyfill.js"></script>
```
are you planning to publish the webnn plyfill package as an npm?

huningxin replied:
It sounds good. Once the polyfill merged, I could that.

huningxin replied:
file issue https://github.com/webmachinelearning/webnn-polyfill/issues/3 to track it.

pyu10055 commented on lenet/lenet.js:
```diff
@@ -0,0 +1,132 @@
+  });
+}
+
+export class Lenet {
```
should this class be named as LeNet?

huningxin replied:
sounds good. I'l fix that.

huningxin replied:
Fixed. Please take a look at the new commit.

pyu10055 reviewed (commented):
THanks Ningxin, this looks good, is it possible to provide a work demo url to try this out?

huningxin commented:
> THanks Ningxin, this looks good, is it possible to provide a work demo url to try this out?

https://huningxin.github.io/webnn-samples/lenet/

huningxin commented:
@anssiko , I fixed the UI issue on narrow screens. I also updated the live version with this fix. Please take a look.

A screenshot:
![image](https://user-images.githubusercontent.com/1005673/92200015-47d23c00-eeab-11ea-9c43-1297de08e0e4.png)

anssiko commented:
Looks great now also on narrow screens, thanks!

pyu10055 commented on lenet/lenet.js:
```diff
@@ -0,0 +1,132 @@
+  async load() {
+    const response = await fetch(this.url_);
+    const arrayBuffer = await response.arrayBuffer();
+    if (arrayBuffer.byteLength !== 1724336) {
```
use a constant for the size

huningxin replied:
good point. I'll fix it.

huningxin replied:
fixed in new commit.

pyu10055 commented on lenet/README.md:
```diff
@@ -0,0 +1,13 @@
+## WebNN API LeNet Example
+The sample uses the LeNet classifications network as an example.
```
It would be good to show a diagram of the LeNet topology with the node names match the code, so readers can follow the code easily.

huningxin replied:
That's a good idea. I'll add it.

huningxin replied:
fixed it by adding a table that lists all nodes. See https://github.com/huningxin/webnn-samples/blob/lenet/lenet/README.md

pyu10055 commented on lenet/README.md:
```diff
@@ -0,0 +1,55 @@
+
+The nodes of the LeNet used in this example are listed in the following table.
+
+| Name | WebNN op | Remarks |
```
can you add the inputs of each op.

huningxin replied:
The model is sequential. I list these ops in topology order. So an op is the input to the op in next row. I can put that notes into the README.md. Does that work?

huningxin replied:
I am thinking it might be better to draw a diagram. I'll come up with one for your review.

huningxin replied:
I added the topology diagram into the README.md. I'am pasting it here for convenience..

![image](https://user-images.githubusercontent.com/1005673/93610654-ead49b00-f9ff-11ea-88e3-2160f7edf1bc.png)

Does it look good to you?

huningxin replied:
After check with @pyu10055 , it would be better to use a higher level illustration. So I use the diagram of [lenet.prototxt](https://github.com/BVLC/caffe/blob/master/examples/mnist/lenet.prototxt) rendered by [netron](https://lutzroeder.github.io/netron/).

![image](https://user-images.githubusercontent.com/1005673/93839040-3d42df80-fcbe-11ea-8136-f1bea1415a3e.png)

I also updated the table that maps the layer to webnn ops.

@pyu10055 , please take another look. Thanks!

pyu10055 replied:
Thanks! This looks great!

huningxin replied:
great, thanks for reviewing!

huningxin commented:
According to [the resolution 01 of WebML CG call on 9/17](https://www.w3.org/2020/09/17-webmachinelearning-minutes.html#r01)

> RESOLUTION: Land WebNN polyfill and samples PRs when existing review comments have been addressed, do not block on in-flight spec PRs and API design discussion

All of the review comments by @pyu10055 have been addressed, so I suppose it is ready to merge. @anssiko , WDYT?

anssiko commented:
We’re ready to land. Thanks!

huningxin commented:
Thanks @anssiko , I am going to merge this PR.
